src/peergos/shared/hamt/Champ.java

```diff
@@ -0,0 +1,524 @@
+        prefix[prefix.length - 1] = new KeyElement(key, val);
+        // ensure canonical structure
+        Arrays.sort(prefix, Comparator.comparing(m -> m.key));
+        result[index] = new HashPrefixPayload(prefix, null);
+
+        return new Champ(dataMap, nodeMap, result);
+    }
+
+    private Champ addNewPrefix(final int bitpos, final ByteArrayWrapper key, final MaybeMultihash val) {
+        final int insertIndex = index(dataMap, bitpos);
+
+        final HashPrefixPayload[] src = this.contents;
+        final HashPrefixPayload[] result = new HashPrefixPayload[src.length + 1];
+
+        System.arraycopy(src, 0, result, 0, insertIndex);
+        System.arraycopy(src, insertIndex, result, insertIndex + 1, src.length - insertIndex);
+        result[insertIndex] = new HashPrefixPayload(new KeyElement[]{new KeyElement(key, val)}, null);
+
+        BitSet newDataMap = BitSet.valueOf(dataMap.toByteArray());
+        newDataMap.set(bitpos);
+        return new Champ(newDataMap, nodeMap, result);
+    }
+
+    private Champ copyAndMigrateFromInlineToNode(final int bitpos, final Pair<Champ, Multihash> node) {
+
+        final int oldIndex = index(dataMap, bitpos);
+        final int newIndex = this.contents.length - 1 - index(nodeMap, bitpos);
+
+        final HashPrefixPayload[] src = this.contents;
+        final HashPrefixPayload[] dst = new HashPrefixPayload[src.length];
+
+        // copy 'src' and remove 1 element at position oldIndex and insert 1 element at position newIndex
+        assert oldIndex <= newIndex;
+        System.arraycopy(src, 0, dst, 0, oldIndex);
+        System.arraycopy(src, oldIndex + 1, dst, oldIndex, newIndex - oldIndex);
+        dst[newIndex] = new HashPrefixPayload(null, MaybeMultihash.of(node.right));
+        System.arraycopy(src, newIndex + 1, dst, newIndex + 1, src.length - newIndex - 1);
+
+        BitSet newNodeMap = BitSet.valueOf(nodeMap.toByteArray());
+        newNodeMap.set(bitpos);
+        BitSet newDataMap = BitSet.valueOf(dataMap.toByteArray());
+        newDataMap.set(bitpos, false);
+        return new Champ(newDataMap, newNodeMap, dst);
+    }
+
+    private Champ overwriteChildLink(final int bitpos, final Pair<Champ, Multihash> node) {
+
+        final int setIndex = this.contents.length - 1 - index(nodeMap, bitpos);
+
+        final HashPrefixPayload[] src = this.contents;
+        final HashPrefixPayload[] dst = new HashPrefixPayload[src.length];
+
+        System.arraycopy(src, 0, dst, 0, src.length);
+        dst[setIndex] = new HashPrefixPayload(null, MaybeMultihash.of(node.right));
+
+        return new Champ(dataMap, nodeMap, dst);
+    }
+
+    public CompletableFuture<Pair<Champ, Multihash>> remove(SigningPrivateKeyAndPublicHash writer,
+                                                            ByteArrayWrapper key,
+                                                            byte[] hash,
+                                                            int depth,
```

Comment:
docstring wherefor art though? Specifically I'd really like to now what depth is.

Reply:
done